zanata-war/src/main/java/org/zanata/action/ProjectHomeAction.java

```diff
@@ -438,4 +489,451 @@ public void resetPageData() {
+        HPerson person = personDAO.findById(data.getPerson().getId());
+        data.setPerson(person);
+        for (PersonProjectMemberships.LocaleRoles roles : data.getLocaleRoles()) {
+            roles.setLocale(localeServiceImpl
+                    .getByLocaleId(roles.getLocale().getLocaleId()));
+        }
+
+        final boolean canManageMembers = identity.hasPermission(project, "manage-members");
+        final boolean canManageTransMembers = identity.hasPermission(project, "manage-translation-members");
+        final boolean canChangeAnyMembers = canManageMembers || canManageTransMembers;
+
+        if (canManageMembers) {
+            project.updateProjectPermissions(data);
+        }
+        if (canManageTransMembers) {
+            project.updateLocalePermissions(data);
+        }
+
+        if (canChangeAnyMembers) {
+            projectDAO.makePersistent(project);
+
+            // Roles may have changed, so role lists are cleared so they will be regenerated
+            personRoles = null;
+            personLocaleRoles = null;
+            project = null;
+
+            // Person may have no roles left and no longer belong in the list, so
+            // ensure the list of people is refreshed.
+            peopleFilterComparator.clearAllMembers();
+            peopleFilterComparator.sortPeopleList();
+        }
+    }
+
+    private final class PeopleFilterComparator extends InMemoryListFilter<HPerson>
+        implements Comparator<HPerson> {
+
+        private final ProjectRolePredicate projectRolePredicate =
+                new ProjectRolePredicate();
+
+        private final ProjectLocalePredicate projectLocalePredicate =
+                new ProjectLocalePredicate();
+
+        private SortingType sortingType;
+
+        @Getter
+        @Setter
+        private boolean showMembersInGroup;
+
+        private List<HPerson> allMembers;
+
+        public PeopleFilterComparator(SortingType sortingType) {
+            this.sortingType = sortingType;
+        }
+
+        @Override
+        public int compare(HPerson o1, HPerson o2) {
+            SortingType.SortOption selectedSortOption =
+                sortingType.getSelectedSortOption();
+
+            if (!selectedSortOption.isAscending()) {
+                HPerson temp = o1;
+                o1 = o2;
+                o2 = temp;
+            }
+
+            if (selectedSortOption.equals(SortingType.SortOption.ROLE)) {
+                setShowMembersInGroup(true);
+            } else {
+                setShowMembersInGroup(false);
+                return o1.getName().toLowerCase()
+                    .compareTo(o2.getName().toLowerCase());
+            }
+
+            // FIXME aeng: this is either confusing, or it is a logic error.
+            //       If the sort option is ROLE, this method will always return 0.
+            //       Please add an explanation of why this is, or fix the logic.
+            return 0;
```

Comment:
Can anyone figure out why this should behave this way? I suspect that maybe people are purposely not sorted when they are grouped, but it does not say so anywhere (and I would expect the people to be sorted alphabetically within groups).